lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp

```diff
@@ -1448,39 +1448,66 @@ void ApplyWorkingDir(SmallString &clang_argument, StringRef cur_working_dir) {
   llvm::sys::path::append(clang_argument, cur_working_dir, rel_path);
   llvm::sys::path::remove_dots(clang_argument);
 }
+
+std::array<StringRef, 2> macro_flags = { "-D", "-U" };
+
+bool IsMultiArgClangFlag(StringRef arg) {
+  for (auto &flag : macro_flags)
+    if (flag == arg)
+      return true;
+  return arg == "-working-directory";
+}
+
+bool IsMacroDefinition(StringRef arg) {
+  for (auto &flag : macro_flags)
+    if (arg.startswith(flag))
+      return true;
+  return false;
+}
+
+bool ShouldUnique(StringRef arg) {
+  return IsMacroDefinition(arg);
+}
 } // namespace
 
 void SwiftASTContext::AddExtraClangArgs(std::vector<std::string> ExtraArgs) {
+  swift::ClangImporterOptions &importer_options = GetClangImporterOptions();
+  llvm::DenseSet<StringRef> unique_flags;
+  for (auto &arg : importer_options.ExtraArgs)
+    unique_flags.insert(arg);
+
   llvm::SmallString<128> cur_working_dir;
   llvm::SmallString<128> clang_argument;
   for (const std::string &arg : ExtraArgs) {
-    // Join multi-arg -D and -U options for uniquing.
+    // Join multi-arg options for uniquing.
     clang_argument += arg;
-    if (clang_argument == "-D" || clang_argument == "-U" ||
-        clang_argument == "-working-directory")
+    if (IsMultiArgClangFlag(clang_argument))
       continue;
 
     auto clear_arg = llvm::make_scope_exit([&] { clang_argument.clear(); });
 
-    // Enable uniquing for -D and -U options.
-    bool is_macro = (clang_argument.size() >= 2 && clang_argument[0] == '-' &&
-                     (clang_argument[1] == 'D' || clang_argument[1] == 'U'));
-    bool unique = is_macro;
-
     // Consume any -working-directory arguments.
     StringRef cwd(clang_argument);
     if (cwd.consume_front("-working-directory")) {
       cur_working_dir = cwd;
       continue;
     }
     // Drop -Werror; it would only cause trouble in the debugger.
-    if (clang_argument.startswith("-Werror")) {
+    if (clang_argument.startswith("-Werror"))
       continue;
-    }
+
+    if (clang_argument.empty())
+      continue;
+
     // Otherwise add the argument to the list.
-    if (!is_macro)
+    if (!IsMacroDefinition(clang_argument))
       ApplyWorkingDir(clang_argument, cur_working_dir);
-    AddClangArgument(clang_argument.str(), unique);
+
+    auto clang_arg_str = clang_argument.str();
+    if (!ShouldUnique(clang_argument) || !unique_flags.count(clang_arg_str)) {
+      importer_options.ExtraArgs.push_back(clang_arg_str);
```

Comment:
@DeeSee doing `importer_options.ExtraArgs.push_back` invalidates all the references inside of `llvm::DenseSet<StringRef> unique_flags`. You might consider using llvm::StringSet.

Reply:
Thanks! I'll make a fix right away.

Just asking — why not `std::unordered_set<std::string>`? (I'm not very familiar with historical causes of why there are custom types for hashmap and hashset in llvm, and using types from std looks less error-prone to me)

Reply:
One (historical) reason is that LLVM predates `std::unordered_set`. However, in this concrete case `llvm::StringSet` will be more efficient than `std::unordered_set<std::string>` because it allocates the strings in the same memory as the set, so you don't have to pay extra memory management overhead for allocating the contents of the individual std::strings.